### PR TITLE
[ALL][CLIENT] Fix serialization for query params specified by ref

### PR DESCRIPTION
**Pull request by sorin-florea**
Fixes https://github.com/OpenAPITools/openapi-generator/issues/907

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (5.3.0), `6.0.x`
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

### CODE REVIEW

**Comment by sorin-florea** [comment]
@wing328

**Comment by sorin-florea** [comment]
@wing328 pinging you again

**Comment by sorin-florea** [comment]
cc https://github.com/orgs/OpenAPITools/teams/generator-core-team

**Comment by guylabs** [comment]
Hi,

Any update on this pull request? Would it be possible to add it to another upcoming `6.0.0` beta release?

Thanks and regards.

**Comment by wing328** [comment]
@sorin-florea thanks for the PR. Please resolve the merge conflicts when you've time.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultGenerator.java [hunk 1]
```diff
@@ -1112,6 +1141,31 @@ private void processOperation(String resourcePath, String httpMethod, Operation
...
+                        parameter.set$ref(schema.get$ref());
+                        schema.set$ref(null);
+                        schema.setType(model.getDataType().toLowerCase(Locale.ROOT));
+
+                        Map<String, Schema> properties = mapProperties(model);
+                        if(!properties.isEmpty()) {
+                            schema.setProperties(properties);
+                        }
+                    }
+                }
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultGenerator.java:1168]: Suggestions: please add some comments explaining what this code block does.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultGenerator.java [hunk 2]
```diff
@@ -1167,6 +1221,63 @@ private void processOperation(String resourcePath, String httpMethod, Operation
...
+                return arraySchema;
+        }
+
+        return null;
+    }
+
+    private BigDecimal toBigDecimal(String value) {
+        return value == null ? null : new BigDecimal(value);
+    }
+
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultGenerator.java:1280]: Suggestions: Add docstrings explaining these new (private) functions

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultGenerator.java (hunk 1 above)
Comment (sorin-florea) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultGenerator.java:1176]: Done

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultGenerator.java (hunk 2 above)
Comment (sorin-florea) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultGenerator.java:1300]: Done

**Comment by wing328** [comment]
@sorin-florea I tested with modules/openapi-generator/src/test/resources/3_0/issue907.yaml and noticed the following changes in openapi.yaml:
```diff
diff --git a/api/openapi.yaml b/api/openapi.yaml
index 998b576..cba8256 100644
--- a/api/openapi.yaml
+++ b/api/openapi.yaml
@@ -17,12 +17,33 @@ paths:
         schema:
           type: string
         style: simple
-      - explode: true
+      - $ref: '#/components/schemas/BuildQuery'
+        explode: true
         in: query
         name: BuildQuery
         required: false
         schema:
-          $ref: '#/components/schemas/BuildQuery'
+          properties:
+            integerProp:
+              format: int32
+              minimum: 0
+              type: integer
+            stringProp:
+              minLength: 1
+              type: string
+            booleanProp:
+              type: boolean
+            arrayProp:
+              items:
+                minLength: 1
+                type: string
+              type: array
+            objectProp:
+              type: object
+            numberProp:
+              minimum: 0
+              type: number
+          type: object
         style: form
       responses:
         "200":
```

In the past few days, we've merged a few PRs to improve the inline model resolver that are doing it the other way around by creating the inline schemas separately and referencing it instead. 

Can we keep $ref as it's instead expanding it into the actual schema?

(I'm working on a PR to improve the inline model resolver so that inline schemas defined in the parameters are created separately and referenced using $ref instead)

I'm free in the afternoon (next couple of hours) so feel free to DM me via Slack to discuss this further.

Thanks again for the PR.

**Comment by wing328** [comment]
One more thing to add is that I don't see changes to the Java classes that fix the serialization for query parameter issue:
```
➜  java-query git:(master) ✗ git status
On branch master
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   .openapi-generator/FILES
	modified:   README.md
	modified:   api/openapi.yaml
	modified:   docs/DefaultApi.md
	modified:   src/main/java/org/openapitools/client/ApiException.java
	modified:   src/main/java/org/openapitools/client/Configuration.java
	modified:   src/main/java/org/openapitools/client/Pair.java
	modified:   src/main/java/org/openapitools/client/StringUtil.java
	modified:   src/main/java/org/openapitools/client/auth/ApiKeyAuth.java
	modified:   src/main/java/org/openapitools/client/auth/HttpBearerAuth.java
	modified:   src/main/java/org/openapitools/client/model/AbstractOpenApiSchema.java
	modified:   src/main/java/org/openapitools/client/model/BuildQuery.java
	modified:   src/main/java/org/openapitools/client/model/SomeReturnValue.java
```

All the changes to the Java files are due to timestamps change, e.g.
```diff
diff --git a/src/main/java/org/openapitools/client/model/BuildQuery.java b/src/main/java/org/openapitools/client/model/BuildQuery.java
index f6fd90d..518a520 100644
--- a/src/main/java/org/openapitools/client/model/BuildQuery.java
+++ b/src/main/java/org/openapitools/client/model/BuildQuery.java
@@ -50,7 +50,7 @@ import org.openapitools.client.JSON;
 /**
  * BuildQuery
  */
-@javax.annotation.Generated(value = "org.openapitools.codegen.languages.JavaClientCodegen", date = "2022-05-13T13:18:02.304890+08:00[Asia/Hong_Kong]")
+@javax.annotation.Generated(value = "org.openapitools.codegen.languages.JavaClientCodegen", date = "2022-05-13T13:21:16.933465+08:00[Asia/Hong_Kong]")
 public class BuildQuery {
   public static final String SERIALIZED_NAME_INTEGER_PROP = "integerProp";
   @SerializedName(SERIALIZED_NAME_INTEGER_PROP)
```

**Comment by wing328** [comment]
FYI. The issue is addressed by https://github.com/OpenAPITools/openapi-generator/pull/12369 instead.

**Comment by newmen** [comment]
As I see, the original PR from @sorin-florea was not merged.
Instead, you referred to #12369, which is clearly not exactly the same as what @sorin-florea did. For example, I don't see the test that was added in #11864 and nothing same in the master.

@wing328 could you please , instead of an incomprehensible link to another PR, somehow summarize the decision on this fix? Which version contains similar one?

**Pull request closed without merging**